controllers/storageconsumer/storageconsumer_controller.go

```diff
@@ -52,11 +73,229 @@ func (r *StorageConsumerReconciler) Reconcile(ctx context.Context, request recon
+	return nil
+}
+
+func (r *StorageConsumerReconciler) reconcileCephClientRbd() error {
+
+	cephClientRBD := r.newCephClientRBD()
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.cephClientRBD, func() error {
+		if err := r.own(r.cephClientRBD); err != nil {
+			return err
+		}
+
+		r.cephClientRBD.Spec = cephClientRBD.Spec
+		return nil
+	})
+
+	if err != nil {
+		r.Log.Error(err, "Failed to update CephClient.", "CephClient", klog.KRef(r.cephClientRBD.Namespace, r.cephClientRBD.Name))
+		return err
+	}
+
+	r.storageConsumer.Status.CephObjects.CephRbdUser = r.cephClientRBD.Name
+	return nil
+}
+
+func (r *StorageConsumerReconciler) newCephBlockPool(failureDomain string) *rookCephv1.CephBlockPool {
+
+	capacity := r.storageConsumer.Spec.Capacity.String()
+
+	return &rookCephv1.CephBlockPool{
+		Spec: rookCephv1.PoolSpec{
+			FailureDomain: failureDomain,
+			Replicated: rookCephv1.ReplicatedSpec{
+				Size:                     3,
+				ReplicasPerFailureDomain: 1,
+				TargetSizeRatio:          .49,
+			},
+			EnableRBDStats: true,
+			Quotas: rookCephv1.QuotaSpec{
+				MaxSize: &capacity,
+			},
+		},
+	}
+}
+
+func (r *StorageConsumerReconciler) newCephClientRBD() *rookCephv1.CephClient {
+
+	return &rookCephv1.CephClient{
+		Spec: rookCephv1.ClientSpec{
+			Caps: map[string]string{
+				"mon": "profile rbd",
+				"osd": fmt.Sprintf("profile rbd pool=%s", r.cephBlockPool.Name),
+			},
+		},
+	}
+}
+
+func (r *StorageConsumerReconciler) areRookCephResourcesReady() bool {
+	// Check status for cephBlockPool resource.
+	if err := r.get(r.cephBlockPool); err == nil {
+		if r.cephBlockPool.Status.Phase != "Ready" {
+			return false
+		}
+	} else {
+		return false
+	}
+
+	// Check status for cephClient resource of rbd.
+	if err := r.get(r.cephClientRBD); err == nil {
+		if r.cephClientRBD.Status.Phase != "Ready" {
+			return false
+		}
+	} else {
+		return false
+	}
+	return true
+}
+
+func (r *StorageConsumerReconciler) get(obj client.Object) error {
+	key := client.ObjectKeyFromObject(obj)
+	return r.Client.Get(r.ctx, key, obj)
+}
+
+func (r *StorageConsumerReconciler) getFailureDomain() (string, error) {
+	storageClusterList := ocsv1.StorageClusterList{}
+	err := r.Client.List(r.ctx, &storageClusterList, client.InNamespace(r.namespace))
+	if err != nil {
+		return "", fmt.Errorf("Error fetching StorageClusterList. %+v", err)
+	}
+	var failureDomain string
+	if len(storageClusterList.Items) == 1 {
+		failureDomain = storageClusterList.Items[0].Status.FailureDomain
+	}
+	return failureDomain, nil
+}
+
+// Checks whether a string is contained within a slice
+func contains(slice []string, s string) bool {
+	for _, item := range slice {
+		if item == s {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *StorageConsumerReconciler) own(resource metav1.Object) error {
+	// Ensure storageConsumer ownership on a resource
+	if err := ctrl.SetControllerReference(r.storageConsumer, resource, r.Scheme); err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
What benefit does this wrapper provide?

Reply:
The wrapper auto-populates the additional info from the reconciler, this way we do not need to do it on every call site. But I would say this can be better  written as:

```suggestion
func (r *StorageConsumerReconciler) own(resource metav1.Object) error {
	// Ensure storageConsumer ownership on a resource
	return ctrl.SetControllerReference(r.storageConsumer, resource, r.Scheme)
}
```